### PR TITLE
zOS: Add 31/64-bit interoperability staging

### PR DESCRIPTION
**Pull request by joransiu**
- Copy the customized JNI header files to include31 directory to JDK_IMAGE base directory.
- Copy the libjvm31.so library and libjvm31.x sidedeck to server / j9vm.

Signed-off-by: Joran Siu <joransiu@ca.ibm.com>

### CODE REVIEW

**Comment by joransiu** [comment]
This change depends on : https://github.com/eclipse-openj9/openj9/pull/13365 to have libjvm31.x and libjvm31.so being built in CMake enabled builds.  It has some additional JNI header and side-deck related changes that is in `openj9-openjdk-jdk11` extensions repo, which is not in this one -- I pulled them along to keep things consistent.   

Once this PR is reviewed, I'll backport to `openj9-openjdk-jdk11` and `openj9-openjdk-jdk17`.

File: closed/custom/Images-pre.gmk
Comment (keithc-ca) [inline closed/custom/Images-pre.gmk:33]: These lines cannot begin with tabs - please use spaces.

Please remove the trailing slashes in the call to `FindFiles` and in the value of `DEST`, and add a comment on the `endif` line.

File: closed/custom/modules/java.base/Copy.gmk
Comment (keithc-ca) [inline closed/custom/modules/java.base/Copy.gmk:39]: Why is 'Interoperability' capitalized?
Please add the missing 'an' in 'part of an OpenJ9 build'.

File: closed/custom/modules/java.base/Copy.gmk
Comment (keithc-ca) [inline closed/custom/modules/java.base/Copy.gmk:41]: Have you verified that using the name `include31` works with `jlink`?

File: closed/custom/modules/java.base/Copy.gmk
Comment (keithc-ca) [inline closed/custom/modules/java.base/Copy.gmk:36]: Please indent lines 35, 36.

File: closed/custom/modules/java.base/Copy.gmk
Comment (keithc-ca) [inline closed/custom/modules/java.base/Copy.gmk:129]: This should be commented:
```
endif # zos
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: closed/custom/modules/java.base/Copy.gmk
Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: Thanks @keithc-ca.  This is a good point -- I hadn't realized jlink images include the headers.  I'll confirm.

Comment (keithc-ca) [reply closed/custom/modules/java.base/Copy.gmk:41]: It's actually the `.jmod` files we are concerned about; they are inputs to `jlink` for producing (possibly custom) images. Some paths are treated specially in `.jmod` files (e.g. `classes/*`); we just need to play by the rules.

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: Right. I see the following in the `java.base.jmod`, not immediately sure how to get additional directories to be included.

```sh
> jar tf java.base.jmod | grep include                                                                                                                                                                     
include/jni.h
include/jvmticmlr.h
include/zos/jni_md.h
include/ibmjvmti.h
include/jvmti.h
include/jni_convert.h
include/classfile_constants.h
```

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: Looks like it's related to this: https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/make/CreateJmods.gmk#L52-L53  though not sure how `include` itself is specified for java.base.

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: Looks like the module specific header files are staged into: ` $(SUPPORT_OUTPUTDIR)/modules_include/$(MODULE)` based on:

https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/make/common/modules/CopyCommon.gmk#L36-L63

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: From what I can tell, jmod appears to support only `--header-files` and will stage files in the given directory into `include` directory of the output jmod file.  Looks like we'd have to modify jmod itself if we want a parallel `include31` directory.  Alternatively, we can have `include31` as a subdirectory within `include` folder by staging the files into modules_include target -- and just document the updated location in migration docs.   There will be another change  to compile /link commands of JNI natives for the libjvm31.x sidedeck file from Java 8, due to lib/s390x/j9vm -> lib/j9vm.   @keithc-ca @pshipton , thoughts?

Comment (keithc-ca) [reply closed/custom/modules/java.base/Copy.gmk:41]: I think it would be simpler to put the new header files in a new subdirectory of `include`, e.g. `include/jni31` or similar. Applications that need to use the 31-bit shim API would compile with `{sdk}/include/jni31` on the include path (instead of `{sdk}/include31`).

Comment (keithc-ca) [reply closed/custom/modules/java.base/Copy.gmk:41]: BTW, a `.jmod` file is not really in in `jar` format; you should use `jmod`:
```
$ ./bin/jmod list jmods/java.base.jmod | grep include | sort
include/classfile_constants.h
include/ibmjvmti.h
include/jni.h
include/jvmticmlr.h
include/jvmti.h
include/linux/jni_md.h
```

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: Thanks @keithc-ca.  I got some initial feedback from 31/64-bit users... no one is against changing it to subdirectory of `include/` for Java 11+.  In some ways, `include/jni31` as you suggested is cleaner.   This simplifies this PR, as none of the `Image-pre.gmk` changes are needed.  I'll touch base with some other users who will be back from Thanksgiving next week to see if anyone raises additional concerns.

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk:41]: Got feedback from users that the suggested change to include/jni31 was acceptable.  I'll update this PR with the necessary changes, and also address the other review comments.  Thanks.

**Comment by joransiu** [comment]
@keithc-ca : I've removed all the Image-Pre.gmk changes, and updated the target directory of the 31-bit JNI headers to include/jni31 of java.base in the latest commit.

File: closed/custom/modules/java.base/Copy.gmk
Comment (keithc-ca) [inline closed/custom/modules/java.base/Copy.gmk:33]: nit: for consistency, the arguments should be reversed (see [OpenJ9.gmk](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/closed/OpenJ9.gmk#L94)):
```
ifeq (zos,$(OPENJDK_TARGET_OS))
```
and the endif line should just be:
```
endif # zos
```
Same goes for the conditional block which starts on line 102.

Comment (joransiu) [reply closed/custom/modules/java.base/Copy.gmk]: Updated under 48c8f92. Thanks!

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Please repeat the second part of ibmruntimes/openj9-openjdk-jdk11#459 to reduce differences between versions.

**Comment by keithc-ca** [comment]
Jenkins compile alinux64 jdknext

**Comment by keithc-ca** [comment]
@pshipton I'll let you merge this when you're satisfied with the changes.

**Review by pshipton — APPROVED** [review]

**Pull request merged**
